packages/sitecore-jss/src/personalize/layout-personalizer.ts

```diff
@@ -34,7 +45,11 @@ function personalizePlaceholder(
   return newComponents;
 }
 
-function personalizeComponent(
+/**
+ * @param {ComponentRenderingWithExpiriences} component component with experiences
+ * @param {string} segment segmentId
+ */
+export function personalizeComponent(
   component: ComponentRenderingWithExpiriences,
   segment: string
 ): ComponentRendering | null {
```

Comment:
Line 67, in order to avoid deep nesting
you can use:
```js
if (!component.placeholders) return;

Object.keys(component.placeholders).forEach((placeholder) => {
  if (!component.placeholders) return;

  component.placeholders[placeholder] = personalizePlaceholder(
    component.placeholders[placeholder],
    segment
  );
});
```